File: small_problems/easy8/leading_substrings.rb
# Write a method that returns a list of all substrings of a string that start at the beginning of the original string. The return value should be arranged in order from shortest to longest substring.

def leading_substrings(string)
  string.chars.map.with_index do |_, index|
    string[0..index]
  end
end

p leading_substrings('abc') == ['a', 'ab', 'abc']
p leading_substrings('a') == ['a']
p leading_substrings('xyzzy') == ['x', 'xy', 'xyz', 'xyzz', 'xyzzy']

File: javascript_small_problems/list_processing/leading_substrings.js
function leadingSubstrings(string) {
  return string.split('').map((_, index) => string.slice(0, index + 1));
}

// Alternate
function leadingSubstrings(string) {
  return [...string].map((_, index) => string.slice(0, index + 1));
}

console.log(leadingSubstrings('abc'));      // ["a", "ab", "abc"]
console.log(leadingSubstrings('a'));        // ["a"]
console.log(leadingSubstrings('xyzzy'));    // ["x", "xy", "xyz", "xyzz", "xyzzy"]